### PR TITLE
Move more private code to public

### PR DESCRIPTION
**Pull request by bumbadadabum**

### CODE REVIEW

File: chat-plugins/chat-monitor.js [hunk 1]
```diff
@@ -6,7 +6,7 @@ const MONITOR_FILE = 'config/chat-plugins/chat-monitor.tsv';
 const WRITE_THROTTLE_TIME = 5 * 60 * 1000;
 
 /** @type {{[k: string]: string[]}} */
-let filterKeys = Chat.filterKeys = Object.assign(Chat.filterKeys, {publicwarn: ['PUBLIC', 'WARN', 'Filtered in public'], warn: ['EVERYWHERE', 'WARN', 'Filtered'], autolock: ['EVERYWHERE', 'AUTOLOCK', 'Autolock'], namefilter: ['NAMES', 'WARN', 'Filtered in usernames'], wordfilter: ['EVERYWHERE', 'FILTERTO', 'Filtered to a different phrase']});
+let filterKeys = Chat.filterKeys = Object.assign(Chat.filterKeys, {publicwarn: ['PUBLIC', 'WARN', 'Filtered in public'], warn: ['EVERYWHERE', 'WARN', 'Filtered'], autolock: ['EVERYWHERE', 'AUTOLOCK', 'Autolock'], namefilter: ['NAMES', 'WARN', 'Filtered in usernames'], wordfilter: ['EVERYWHERE', 'FILTERTO', 'Filtered to a different phrase'], shorteners: ['EVERYWHERE', 'SHORTENER', 'Url Shorteners']});
```
Comment (Zarel) [inline chat-plugins/chat-monitor.js]: I'm trying to enforce a line length limit right now.

```suggestion
let filterKeys = Chat.filterKeys = Object.assign(Chat.filterKeys, {
	publicwarn: ['PUBLIC', 'WARN', 'Filtered in public'],
	warn: ['EVERYWHERE', 'WARN', 'Filtered'],
	autolock: ['EVERYWHERE', 'AUTOLOCK', 'Autolock'],
	namefilter: ['NAMES', 'WARN', 'Filtered in usernames'],
	wordfilter: ['EVERYWHERE', 'FILTERTO', 'Filtered to a different phrase'],
	shorteners: ['EVERYWHERE', 'SHORTENER', 'Url Shorteners'],
});
```

**Comment by panpawn** [comment]
Maybe this should also be apart of ``/roomsettings``?

File: chat-plugins/chat-monitor.js (hunk 1 above)
Comment (bumbadadabum) [reply chat-plugins/chat-monitor.js]: done

**Comment by Zarel** [comment]
This pull request **introduces 2 alerts** when merging c5963a352eddfaf64377ede0c2e3aeecc8113f56 into c934c04bbcecd377671c02581ea10dfdaf359d29 - [view on LGTM.com](https://lgtm.com/projects/g/Zarel/Pokemon-Showdown/rev/pr-961a26479168ae5b25232575677abebb6c4e2e72)

**new alerts:**

* 2 for Incomplete string escaping or encoding

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by bumbadadabum** [comment]
> This pull request **introduces 2 alerts** when merging [c5963a3](https://github.com/Zarel/Pokemon-Showdown/commit/c5963a352eddfaf64377ede0c2e3aeecc8113f56) into [c934c04](https://github.com/Zarel/Pokemon-Showdown/commit/c934c04bbcecd377671c02581ea10dfdaf359d29) - [view on LGTM.com](https://lgtm.com/projects/g/Zarel/Pokemon-Showdown/rev/pr-961a26479168ae5b25232575677abebb6c4e2e72)
> 
> **new alerts:**
> 
>     * 2 for Incomplete string escaping or encoding
> 
> 
> _Comment posted by [LGTM.com](https://lgtm.com)_

I copied this replace directly from your code so idk what's wrong with it. It also works fine.

**Comment by Zarel** [comment]
I mean, my code is nonzero likely to have bugs. Just because LGTM couldn't see them before doesn't mean they don't exist.

File: chat-plugins/chat-monitor.js [hunk 2]
```diff
@@ -34,11 +41,12 @@ setImmediate(() => {
 					if (punishment === 'FILTERTO') {
 						const filterTo = rest[0];
 						filterWords[key].push([new RegExp(word, 'ig'), reason, filterTo, parseInt(times) || 0]);
-						continue loop;
+					} else if (punishment === 'SHORTENER') {
+						filterWords[key].push([new RegExp(`\\b${word.replace(/\./g, '\\.')}\\b`), reason, null, parseInt(times) || 0]);
```
Comment (Zarel) [inline chat-plugins/chat-monitor.js]: LGTM:
> This does not backslash-escape the backslash character.

File: chat-plugins/chat-monitor.js [hunk 3]
```diff
@@ -331,8 +358,13 @@ let commands = {
 				let [word, ...reasonParts] = rest;
 				word = word.trim();
 				let reason = reasonParts.join(',').trim();
-				if (filterWords[list].some(val => val[0] === word)) return this.errorReply(`${word} is already added to the ${list} list.`);
-				filterWords[list].push([word, reason, null, 0]);
+				if (filterKeys[list][1] === 'SHORTENER') {
+					if (filterWords[list].some(val => String(val[0]).slice(3, -3).replace(/\\./g, '.') === word)) return this.errorReply(`${word} is already added to the ${list} list.`);
+					filterWords[list].push([new RegExp(`\\b${word.replace(/\./g, '\\.')}\\b`), reason, null, 0]);
```
Comment (Zarel) [inline chat-plugins/chat-monitor.js]: LGTM
> This does not backslash-escape the backslash character.

**Comment by Zarel** [comment]
This pull request **introduces 2 alerts** when merging 1eaf7f8cbe54e0ee27ac3f9b7fd0376679ca8001 into e73c6f367da09fb2c3816f80fa738c3042ea920b - [view on LGTM.com](https://lgtm.com/projects/g/Zarel/Pokemon-Showdown/rev/pr-c9dd6d4c8d5ab6a79226f4d58b58726895e110a7)

**new alerts:**

* 2 for Incomplete string escaping or encoding

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by Zarel** [comment]
This pull request **introduces 2 alerts** when merging 371dd63e12da16403bc59f1f14da0b20de085a73 into e73c6f367da09fb2c3816f80fa738c3042ea920b - [view on LGTM.com](https://lgtm.com/projects/g/Zarel/Pokemon-Showdown/rev/pr-0c3ae4ea5f756af811358ee198c1479467ebc96a)

**new alerts:**

* 2 for Incomplete string escaping or encoding

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by Zarel** [comment]
https://lgtm.com/help/lgtm/alert-suppression

Their examples show no space.

**Comment by xcorail** [comment]
Hello @Zarel @bumbadadabum 

Sorry for this, I understand your frustration. 
Currently, the automated code review mechanism doesn't take into account the suppression comments. However, the analysis do: if you merge this PR, you'll see that the alerts disappear from the alerts page.

We are currently discussing internally about this, and in the meantime I will ask to update the documentation to reflect this behaviour.

Regards

**Comment by Zarel** [comment]
Honestly, you could just completely sanitize it. :|

File: chat-plugins/chat-monitor.js [hunk 4]
```diff
@@ -34,11 +41,12 @@ setImmediate(() => {
 					if (punishment === 'FILTERTO') {
 						const filterTo = rest[0];
 						filterWords[key].push([new RegExp(word, 'ig'), reason, filterTo, parseInt(times) || 0]);
-						continue loop;
+					} else if (punishment === 'SHORTENER') {
+						filterWords[key].push([new RegExp(`\\b${word.replace(/\./g, '\\.')}\\b`), reason, null, parseInt(times) || 0]); // lgtm[js/incomplete-sanitization]
```
Comment (Zarel) [inline chat-plugins/chat-monitor.js]:
```suggestion
						const regex = new RegExp(`\\b${word.replace(/\\/g, '\\\\').replace(/\./g, '\\.')}\\b`);
						filterWords[key].push([regex, reason, null, parseInt(times) || 0]);
```

File: chat-plugins/chat-monitor.js [hunk 5]
```diff
@@ -331,8 +358,13 @@ let commands = {
 				let [word, ...reasonParts] = rest;
 				word = word.trim();
 				let reason = reasonParts.join(',').trim();
-				if (filterWords[list].some(val => val[0] === word)) return this.errorReply(`${word} is already added to the ${list} list.`);
-				filterWords[list].push([word, reason, null, 0]);
+				if (filterKeys[list][1] === 'SHORTENER') {
+					if (filterWords[list].some(val => String(val[0]).slice(3, -3).replace(/\\./g, '.') === word)) return this.errorReply(`${word} is already added to the ${list} list.`);
+					filterWords[list].push([new RegExp(`\\b${word.replace(/\./g, '\\.')}\\b`), reason, null, 0]); // lgtm[js/incomplete-sanitization]
```
Comment (Zarel) [inline chat-plugins/chat-monitor.js]:
```suggestion
					const regex = new RegExp(`\\b${word.replace(/\\/g, '\\\\').replace(/\./g, '\\.')}\\b`)
					filterWords[list].push([regex, reason, null, 0]);
```

File: chat-plugins/chat-monitor.js [hunk 6]
```diff
@@ -114,6 +129,17 @@ let chatfilter = function (message, user, room) {
 			return false;
 		}
 	}
+	for (let i = 0; i < filterWords.shorteners.length; i++) {
+		let [regex] = filterWords.shorteners[i];
+		if (typeof regex === 'string') continue; // Failsafe to appease typescript.
+		if (regex.test(lcMessage)) {
+			if ((room && ((room.chatRoomData && room.id.endsWith('staff')) || room.id.startsWith('help-'))) || user.isStaff || (this.pmTarget && this.pmTarget.isStaff)) return `${message} __[shortener: ${String(regex).slice(3, -3).replace('\\.', '.')}]__`;
```
Comment (Zarel) [inline chat-plugins/chat-monitor.js]:
```suggestion
			if ((room && ((room.chatRoomData && room.id.endsWith('staff')) || room.id.startsWith('help-'))) || user.isStaff || (this.pmTarget && this.pmTarget.isStaff)) {
				return `${message} __[shortener: ${String(regex).slice(3, -3).replace('\\.', '.')}]__`;
			}
```

File: chat-plugins/chat-monitor.js [hunk 7]
```diff
@@ -114,6 +129,17 @@ let chatfilter = function (message, user, room) {
 			return false;
 		}
 	}
+	for (let i = 0; i < filterWords.shorteners.length; i++) {
+		let [regex] = filterWords.shorteners[i];
+		if (typeof regex === 'string') continue; // Failsafe to appease typescript.
```
Comment (Zarel) [inline chat-plugins/chat-monitor.js]: This is a fail-dangerous, not a failsafe. You should `throw` if this is a situation you think will never happen.

Comment (Zarel) [reply chat-plugins/chat-monitor.js]:
```suggestion
		if (typeof regex === 'string') throw new Error(`shortener filters should not have strings`);
```

File: chat-plugins/chat-monitor.js (hunk 6 above)
Comment (Zarel) [reply chat-plugins/chat-monitor.js]: Honestly I'll need to refactor this anyway; I guess I'll just make it a separate commit. Have fun dealing with the merge conflict.

Comment (Zarel) [reply chat-plugins/chat-monitor.js]:
```suggestion
			if (isStaff) return `${message} __[shortener: ${String(regex).slice(3, -3).replace('\\.', '.')}]__`;
```

**Comment by bumbadadabum** [comment]
Since Zarel said I can merge this myself.

**Pull request merged**